### PR TITLE
Add GLM-4 CPU example

### PR DESCRIPTION
**Pull request by JinBridger**

### CODE REVIEW

File: python/llm/example/CPU/HF-Transformers-AutoModels/Model/glm4/README.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+We suggest using conda to manage environment:
+
+On Linux:
+
+```bash
+conda create -n llm python=3.11 # recommend to use Python 3.11
+conda activate llm
+
+# install the latest ipex-llm nightly build with 'all' option
+pip install --pre --upgrade ipex-llm[all] --extra-index-url https://download.pytorch.org/whl/cpu
```
Comment (Oscilloscope98) [inline python/llm/example/CPU/HF-Transformers-AutoModels/Model/glm4/README.md:20]: It seems like `tiktoken` is a required package for GLM-4?

**Comment by JinBridger** [comment]
```
(jinqiao-llm-0612) arda@arda-arc21:~/jinqiao$ python generate.py --repo-id-or-model-path /mnt/disk1/models/glm-4-9b-chat/
Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:00<00:00, 41.15it/s]
2024-06-12 09:59:29,390 - INFO - Converting the current model to sym_int4 format......
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
Traceback (most recent call last):
  File "/home/arda/jinqiao/generate.py", line 61, in <module>
    output = model.generate(input_ids,
             ^^^^^^^^^^^^^^^^^^^^^^^^^
......
  File "/home/arda/miniforge3/envs/jinqiao-llm-0612/lib/python3.11/site-packages/ipex_llm/transformers/models/chatglm4.py", line 342, in core_attn_forward
    key_layer.transpose(2, 3)) / math.sqrt(head_dim)
                                 ^^^^
NameError: name 'math' is not defined
```

Looks like we need `import math` in [transformers/models/chatglm4.py](https://github.com/intel-analytics/ipex-llm/blob/main/python/llm/src/ipex_llm/transformers/models/chatglm4.py) ?

File: python/llm/example/CPU/HF-Transformers-AutoModels/Model/glm4/README.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
+# GLM-4
```
Comment (Oscilloscope98) [inline python/llm/example/CPU/HF-Transformers-AutoModels/Model/glm4/README.md:1]: Missing link update in tables

File: python/llm/example/CPU/HF-Transformers-AutoModels/Model/glm4/README.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+# set IPEX-LLM env variables
+source ipex-llm-init
+
+# e.g. for a server with 48 cores per socket
+export OMP_NUM_THREADS=48
+numactl -C 0-47 -m 0 python ./generate.py
+```
+
+#### 2.3 Sample Output
+#### [THUDM/glm-4-9b-chat](https://huggingface.co/THUDM/glm-4-9b-chat)
```
Comment (Oscilloscope98) [inline python/llm/example/CPU/HF-Transformers-AutoModels/Model/glm4/README.md:68]: `####` -> `#####`

File: python/llm/example/CPU/HF-Transformers-AutoModels/Model/glm4/README.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+### 1. Install
+We suggest using conda to manage environment:
+
+On Linux:
+
+```bash
+conda create -n llm python=3.11 # recommend to use Python 3.11
+conda activate llm
+
+# install the latest ipex-llm nightly build with 'all' option
```
Comment (Oscilloscope98) [inline python/llm/example/CPU/HF-Transformers-AutoModels/Model/glm4/README.md:19]: Make it the same as we do for GPU: `# install tiktoken required for GLM-4`

File: python/llm/example/CPU/PyTorch-Models/Model/glm4/README.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+#### 2.4 Sample Output
+#### [THUDM/glm-4-9b-chat](https://huggingface.co/THUDM/glm-4-9b-chat)
+```log
+Inference time: xxxx s
+-------------------- Output --------------------
+
+AI是什么？
+
+AI，即人工智能（Artificial Intelligence），是指由人创造出来的，能够模拟、延伸和扩展人的智能的计算机系统或机器。人工智能技术
+```
```
Comment (Oscilloscope98) [inline python/llm/example/CPU/PyTorch-Models/Model/glm4/README.md:75]: Missing sample output in english

**Comment by Oscilloscope98** [comment]
> ```
> (jinqiao-llm-0612) arda@arda-arc21:~/jinqiao$ python generate.py --repo-id-or-model-path /mnt/disk1/models/glm-4-9b-chat/
> Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:00<00:00, 41.15it/s]
> 2024-06-12 09:59:29,390 - INFO - Converting the current model to sym_int4 format......
> Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
> Traceback (most recent call last):
>   File "/home/arda/jinqiao/generate.py", line 61, in <module>
>     output = model.generate(input_ids,
>              ^^^^^^^^^^^^^^^^^^^^^^^^^
> ......
>   File "/home/arda/miniforge3/envs/jinqiao-llm-0612/lib/python3.11/site-packages/ipex_llm/transformers/models/chatglm4.py", line 342, in core_attn_forward
>     key_layer.transpose(2, 3)) / math.sqrt(head_dim)
>                                  ^^^^
> NameError: name 'math' is not defined
> ```
> 
> Looks like we need `import math` in [transformers/models/chatglm4.py](https://github.com/intel-analytics/ipex-llm/blob/main/python/llm/src/ipex_llm/transformers/models/chatglm4.py) ?

It will be fixed in https://github.com/intel-analytics/ipex-llm/pull/11276. We may use 2.1.0b20240606 for sample output now :)

File: python/llm/example/CPU/HF-Transformers-AutoModels/Model/glm4/generate.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+                        help='Prompt to infer')
+    parser.add_argument('--n-predict', type=int, default=32,
+                        help='Max tokens to predict')
+
+    args = parser.parse_args()
+    model_path = args.repo_id_or_model_path
+
+    # Load model in 4 bit,
+    # which convert the relevant layers in the model into INT4 format
+    model = AutoModel.from_pretrained(model_path,
```
Comment (Oscilloscope98) [inline python/llm/example/CPU/HF-Transformers-AutoModels/Model/glm4/generate.py:44]: Missing some keyword arguments when loading the model

File: python/llm/example/CPU/HF-Transformers-AutoModels/Model/glm4/generate.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    
+    # Generate predicted tokens
+    with torch.inference_mode():
+        prompt = GLM4_PROMPT_FORMAT.format(prompt=args.prompt)
+        input_ids = tokenizer.encode(prompt, return_tensors="pt")
+        st = time.time()
+        # if your selected model is capable of utilizing previous key/value attentions
+        # to enhance decoding speed, but has `"use_cache": false` in its model config,
+        # it is important to set `use_cache=True` explicitly in the `generate` function
+        # to obtain optimal performance with IPEX-LLM INT4 optimizations
```
Comment (Oscilloscope98) [inline python/llm/example/CPU/HF-Transformers-AutoModels/Model/glm4/generate.py:60]: We could remove the comments for use_cache here if we have it when loading the model :)

File: python/llm/example/CPU/HF-Transformers-AutoModels/Model/glm4/README.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+On Windows:
+
+```cmd
+conda create -n llm python=3.11
+conda activate llm
+
+pip install --pre --upgrade ipex-llm[all]
+
+# install tiktoken required for GLM-4
```
Comment (Oscilloscope98) [inline python/llm/example/CPU/HF-Transformers-AutoModels/Model/glm4/README.md:125]: Remove `# install tiktoken required for GLM-4` here. Not a comment format for cmd :)

File: python/llm/example/CPU/PyTorch-Models/Model/glm4/README.md [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+#### 2.3 Arguments Info
+In the example, several arguments can be passed to satisfy your requirements:
+
+- `--repo-id-or-model-path`: str, argument defining the huggingface repo id for the GLM-4 model to be downloaded, or the path to the huggingface checkpoint folder. It is default to be `'THUDM/glm-4-9b-chat'`.
+- `--prompt`: str, argument defining the prompt to be inferred (with integrated prompt format for chat). It is default to be `'AI是什么？'`.
+- `--n-predict`: int, argument defining the max number of tokens to predict. It is default to be `32`.
+
+#### 2.4 Sample Output
+#### [THUDM/glm-4-9b-chat](https://huggingface.co/THUDM/glm-4-9b-chat)
```
Comment (Oscilloscope98) [inline python/llm/example/CPU/PyTorch-Models/Model/glm4/README.md:71]: `####` -> `#####`

**Review by Oscilloscope98 — APPROVED** [review]

**Pull request merged**